File: ruby/gogen.rb
class Gogen
  TEXT_INPUT_SPLIT = "#####"
  GRID_SIZE = 5
  PRINT_PADDING = 3
  BLANK_CHARACTER = '_'


  def initialize(input)
    initialize_from_text(input) if input.is_a?(String)
  end

  def initialize_from_text(input)
    parts = input.split(TEXT_INPUT_SPLIT)
    # transforms V _ D _ J\n_ _ _ _ _\nQ _ R _ S\n_ _ _ _ _\nX _ H _ M\n"
    # into grid style system, with x and y row and col indexes
    @grid = parts[0].split("\n").map(&:split).transpose
    @words = parts[1].split("\n").reject(&:empty?)
    # find used letters and remaining letters
    @letters_found = build_found_letters_with_pos(@grid)
    @letters_to_find = build_unfound_letter_with_remaining_pos(@letters_found)
    @adjacencies = build_adjacencies(@words)
  end

  # returns hash of letter and current co-ords
  def build_found_letters_with_pos(grid)
    letters = Hash.new { |h, k| h[k] = {} }
    @grid.each_with_index do |col, col_index|
      col.each_with_index do |letter, row_index|
        letters[letter] = [col_index, row_index] unless letter.eql?(BLANK_CHARACTER)
      end
    end
    letters
  end

  def build_unfound_letter_with_remaining_pos(letters_found)
    letters_to_find = ('A'..'Y').to_a - letters_found.keys
    all_positions = (0..GRID_SIZE-1).to_a.product((0..GRID_SIZE-1).to_a)
    positions_to_find = all_positions - letters_found.values
    letters_to_find.reduce({}) do |memo, letter|
      memo[letter] = positions_to_find
      memo
    end
  end

  def build_neighbourhood(pos)
    x, y = pos
    x_range = ([0, x-1].max..[GRID_SIZE-1, x+1].min).to_a
    y_range = ([0, y-1].max..[GRID_SIZE-1, y+1].min).to_a
    x_range.product(y_range)
  end

  def solve!
    log "Let's solve a Gogen Puzzle!"

    while unsolved?
      @letters_to_find.each_key do |letter|
        @adjacencies[letter].each do |adjacent_letter|
          # get current possible (or known single) position of letter
          positions_of_letter = @letters_to_find[adjacent_letter] || [@letters_found[adjacent_letter]]
          # build neighhbourhoods and union
          valid_positions = positions_of_letter.map do |pos|
            build_neighbourhood(pos)
          end.flatten(1).uniq
          # reject known fixed positions
          valid_positions = valid_positions - @letters_found.values
          # update possible positions with intersection
          @letters_to_find[letter] = @letters_to_find[letter] & valid_positions
          if @letters_to_find[letter].size == 1
            @letters_found[letter] = @letters_to_find[letter].first
            @letters_to_find.delete(letter)
            break # found it, next loop
          end
        end
      end
    end

    # lets reapply to the grid
    @letters_found.each do |letter, pos|
      @grid[pos[0]][pos[1]] = letter
    end
  end

  # build a list of what letters are next to each letter in the word
  def build_adjacencies(words)
    adjacencies = Hash.new { |h, k| h[k] = [] }
    words.each do |word|
      (word.size-1).times do |i|
        first = word[i]
        second = word[i+1]
        adjacencies[first].push(second)
        adjacencies[second].push(first)
      end
    end
    adjacencies
  end

  def print!
    PRINT_PADDING.times { puts "\n" }
    @grid.each do |row|
      puts ["\t" * PRINT_PADDING, row.join("  ")].flatten.join("")
      puts "\n"
    end
    PRINT_PADDING.times { puts "\n" }
  end

  def verify_solved!
    solved = @words.map do |word|
      found = 1
      (word.size - 1).times do |i|
        pos = @letters_found[word[i]]
        break if pos.nil?
        build_neighbourhood(pos).each do |neighbour_pos|
          if @grid[neighbour_pos[0]][neighbour_pos[1]] == word[i+1]
            found += 1
            break
          end
        end
      end
      found.eql?(word.size)
    end.all? { true }
    log "Actually Solved? #{solved}"
  end

  def unsolved?
    @letters_to_find.any?
  end

  def log(msg)
    puts "[GOGEN] #{msg}"
  end
end

PUZZLE = ENV['PUZZLE'] || 1
path = File.join(File.dirname(__FILE__), ["..", "examples", "#{PUZZLE}-unsolved.txt"])
puzzle = Gogen.new(File.read(path))

puzzle.solve!
puzzle.print!
puzzle.verify_solved!

# fix up ? use sets rather than arrays (if necesary in ruby)
# work out big O time - f(grid_size = G, neighhbourhood size = 8, # of unfound letters = L, alphabet = A)
# interestingly how it breaks if edge fixed are removed, but middle R can be removed

File: go/gogen.go
package main

import (
  "fmt"
  "io/ioutil"
  "path/filepath"
  "os"
  "strings"
  set "github.com/deckarep/golang-set"
)

const textInputSplit = "#####"
const gridSize = 5
const blankCharacter = "_"
const availableLetters = "ABCDEFGHIJKLMNOPQRSTUVWXY"

type Pos struct {
  x int
  y int
}

type Grid [gridSize][gridSize]string
type WordList []string
type LetterPosSet map[string]set.Set
type AdjacencyMap map[string]set.Set

func log(msg string) {
  fmt.Printf("[GOGEN] %s\n", msg)
}

func checkErr(e error) {
  if e != nil {
    panic(e)
  }
}

func intMin(x, y int) int {
    if x < y {
        return x
    }
    return y
}

func intMax(x, y int) int {
    if x > y {
        return x
    }
    return y
}


func getLetterPosSetKeys(input LetterPosSet) (set.Set) {
  keys := set.NewSet()
  for k := range(input) {
    keys.Add(k)
  }
  return keys
}

func textInputToGrid(textInput string) (Grid, LetterPosSet) {
  newGrid := [gridSize][gridSize]string{}
  lettersFound := make(LetterPosSet)
  lines := strings.Split(strings.Trim(textInput, "\n"), "\n")
  for row := range(lines) {
    cells := strings.Split(lines[row], " ")
    for col, val := range(cells) {
      newGrid[col][row] = val
      posSet := set.NewSet(Pos{col, row})
      if (val != blankCharacter) {
        lettersFound[val] = posSet
      }
    }
  }
  return newGrid, lettersFound
}

func extractDataFromFile(puzzle string) (Grid, LetterPosSet, WordList) {
  absPath, err := filepath.Abs(fmt.Sprintf("../examples/%s-unsolved.txt", puzzle))
  checkErr(err)
  data, err := ioutil.ReadFile(absPath)
  checkErr(err)
  parts := strings.Split(string(data), textInputSplit)
  grid, lettersFound := textInputToGrid(parts[0])
  words := strings.Split(parts[1], "\n")
  return grid, lettersFound, words
}

func buildAllPositions() (set.Set) {
  s := set.NewSet()
  for row := 0; row < gridSize; row++ {
    for col := 0; col < gridSize; col++ {
      s.Add(Pos{col, row})
    }
  }
  return s
}

func buildLettersToFind(lettersFound LetterPosSet) (LetterPosSet) {
  allPositions := buildAllPositions()
  allLetters := strings.Split(availableLetters, "")
  allLettersSet := set.NewSet()
  for _, l := range(allLetters) {
    allLettersSet.Add(l)
  }

  knownLetters := getLetterPosSetKeys(lettersFound)
  knownPositions := set.NewSet()
  for _, ps := range(lettersFound) {
    ps.Each(func(elem interface{}) bool {
      knownPositions.Add(elem)
      return false
    })
  }

  missingLetters := allLettersSet.Difference(knownLetters)
  blankPositions := allPositions.Difference(knownPositions)

  lettersToFind := make(LetterPosSet)
  missingLetters.Each(func(letter interface{}) bool {
    l, _ := letter.(string)
    blankPositionCopy := blankPositions.Clone()
    lettersToFind[l] = blankPositionCopy   // is there any easier way to copy a set?
    return false
  })
  return lettersToFind
}

func buildAdjacencies(words WordList) (AdjacencyMap) {
  adjMap := make(AdjacencyMap)
  for _, word := range(words) {
    chars := strings.Split(word, "")
    maxLoop := len(chars) - 1;
    for i := 0; i < maxLoop; i++ {
      firstChar := chars[i]
      secondChar := chars[i+1]
      if (adjMap[firstChar] == nil) { adjMap[firstChar] = set.NewSet() }
      if (adjMap[secondChar] == nil) { adjMap[secondChar] = set.NewSet() }
      adjMap[firstChar].Add(secondChar)
      adjMap[secondChar].Add(firstChar)
    }
  }
  return adjMap
}

// done use math.Min https://mrekucci.blogspot.co.uk/2015/07/dont-abuse-mathmax-mathmin.html
func buildNeighbourhood(pos Pos) (set.Set) {
  s := set.NewSet()
  x_min := intMax(pos.x - 1, 0)
  y_min := intMax(pos.y - 1, 0)
  x_max := intMin(pos.x + 1, gridSize - 1)
  y_max := intMin(pos.y + 1, gridSize - 1)
  for x := x_min; x <= x_max; x++ {
    for y := y_min; y <= y_max; y++ {
      s.Add(Pos{x, y})
    }
  }
  return s
}


func solve(lettersToFind LetterPosSet, lettersFound LetterPosSet, adjacencies AdjacencyMap, grid Grid) (Grid) {
  // continue to iterate until we have all letters, this map
  // will get updated further down
  for len(lettersToFind) > 0 {
    for letter, _ := range(lettersToFind) {
      adjacencies[letter].Each(func(aj interface{}) bool {
        adjacentLetter, _ := aj.(string)
        positionsOfLetter := set.NewSet() // in both cases lets make this a set
        if (lettersFound[adjacentLetter] != nil) {
          // we found this already so we know where it is, only one possible position
          positionsOfLetter = lettersFound[adjacentLetter].Clone()
        } else {
          // not found so lets get all the possible positions
          positionsOfLetter = lettersToFind[adjacentLetter].Clone()
        }
        // now we work out all possible positions of that letter, build
        // neighhbourhoodas for each and flatten/uniq by using union
        validPositions := set.NewSet()
        positionsOfLetter.Each(func(ps interface{}) bool {
          pos := ps.(Pos)
          validPositions = validPositions.Union(buildNeighbourhood(pos))
          return false
        })
        // now remove all position we already now
        knownPositions := set.NewSet()
        for _, ps := range(lettersFound) {
          ps.Each(func(elem interface{}) bool {
            knownPositions.Add(elem)
            return false
          })
        }
        validPositions = validPositions.Difference(knownPositions)

        // now update the original lettersToFind by intersect with our hopefully reduced list
        lettersToFind[letter] = lettersToFind[letter].Intersect(validPositions)

        if (lettersToFind[letter].Cardinality() == 1) {
          lettersFound[letter] = lettersToFind[letter]
          delete(lettersToFind, letter)
          return true // breaks out of the inner Each, cant break in callback Each rather than for loop
        } else {
          return false
        }
      })
    }
  }

  for letter, positions := range(lettersFound) {
    // should only be one position but we have to user Iter to get first element out
    for ps := range(positions.Iter()) {
      pos := ps.(Pos)
      grid[pos.x][pos.y] = letter
    }
  }

  return grid;
}

func printGrid(grid Grid) {
  for i := range(grid) {
    fmt.Printf("\t\t%s\n", strings.Join(grid[i][:], " "))
  }
}

func main() {
  puzzle := "1"
  if len(os.Args) > 1 {
    puzzle = os.Args[1]
  }
  grid, lettersFound, words := extractDataFromFile(puzzle);
  printGrid(grid)
  log(fmt.Sprintf("Lets solve a gogen puzzle %s!\n", puzzle))
  updatedGrid := solve(buildLettersToFind(lettersFound), lettersFound, buildAdjacencies(words), grid)
  printGrid(updatedGrid)
}
